Add ScyllaDB driver support

porunov opened the pull request:
- Add ScyllaDB support via `janusgraph-scylla`
- `janusgraph-cql` contains tests for Cassandra3, Cassandra4, ScyllaDB. `janusgraph-scylla` contains tests for ScyllaDB only.

ScyllaDB driver has the same classpath as DataStax driver. Thus, we can't have both `janusgraph-cql` and `janusgraph-scylla` dependencies in the classpath at the same time. 
It means that `janusgraph-all` can have only one of them. This is fine I think because anyone who is using that dependency should be able to exclude `janusgraph-cql` and include `janusgraph-scylla` when necessary.

That said, it also means that we can't have `janusgraph-scylla` in any of the JanusGraph distributions because those distributions already have `janusgraph-cql` dependency. Right now the only option to add `scylla` as a storage adapter to JanusGraph distribution is by manually replacing libraries. 
The issue to make it more easier for users to replace `cql` drivers with `scylla` drivers is here: #3580 (has a solution description).

I checked that the optimized ScyllaDB driver is used for `scylla` storage option.

Fixes #1778
Related to #2451 
Fixes to #2505 (comment about observations of Hadoop tests with ScyllaDB is [here](https://github.com/JanusGraph/janusgraph/issues/2505#issuecomment-1427215950))

Signed-off-by: Oleksandr Porunov <alexandr.porunov@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

porunov commented:
If anyone knows why jacoco complains when `-Pcovarage` is used for ScyllaDB tests, please let me know because I couldn't figure out the problem yet.
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:3.1.0:run (default-cli) on project janusgraph-scylla: An Ant BuildException has occured: Unable to read execution data file /home/runner/work/janusgraph/janusgraph/target/jacoco-combined.exec
Error:  around Ant part ...<report>... @ 5:13 in /home/runner/work/janusgraph/janusgraph/janusgraph-scylla/target/antrun/build-main.xml
```
If I omit `-Pcovarage` parameter then it builds fine.

farodin91 commented:
@porunov  Would it be possible to use java shadowing for instead of library coping?

porunov commented:
> @porunov Would it be possible to use java shadowing for instead of library coping?

I don't have experience with shading Java libraries, so can't say for sure. I have created an issue #3580 to be able to avoid copying libraries. In that issue I proposed making diff between `cql` and `scylla` dependencies, split them to multiple folders and choose the folder to include based on environment variable. 
I think shading should be another approach to solve #3580 but I don't have a clear picture of how to accomplish it with shading due to lack of experience.

FlorianHockmann commented on docs/changelog.md:
```diff
@@ -236,6 +237,18 @@ ElasticSearch 8 doesn't allow creating new indexes with `Mapping.PREFIX_TREE` ma
 using `Mapping.PREFIX_TREE` will work in ElasticSearch 8 after migration. See 
 [ElasticSearch 8 migration guide](https://www.elastic.co/guide/en/elasticsearch/reference/current/migrating-8.0.html#geo-shape-strategy).
 
+##### Add support for ScyllaDB driver
```
Not that important, but I would move this section and the one above about ES 8 at least above the deprecation notices. Those deprecations should be the last part of the changelog and I would order the changelog in general by priority. Maybe move both sections right behind the breaking changes?

porunov replied:
Moving both sections above the deprecation notices, but we probably need to reorder that list before the release.

FlorianHockmann commented on docs/storage-backend/scylladb.md:
```diff
@@ -27,3 +27,69 @@ ScyllaDB is fully compatible with Cassandra. To use it as the data storage layer
+
+!!! note
+    The manual process described below is temporary and should be replaced by automated process after the issue
+    [janusgraph/janusgraph#3580](https://github.com/JanusGraph/janusgraph/issues/3580) is resolved.
```
Would be really great if we could get this resolved until the 1.0.0 release.

porunov replied:
Agree. I don't really want users to handle manual copy/paste process. I think we should just change the distribution build process to have `lib`, `cqlLib`, `scyllaLib` folders as noted in #3580 
It should be fairly simple for users to switch between CQL or Scylla drivers by adding an optional environment variable. 
@farodin91 also [mentioned](https://github.com/JanusGraph/janusgraph/pull/3578#issuecomment-1428032096) using Java Shadowing approach which could potentially be an alternative version to resolve #3580 , but I haven't had a chance to investigate the approach with Java Shadowing.

FlorianHockmann commented on janusgraph-scylla/src/main/java/org/janusgraph/diskstorage/cql/ScyllaStoreManager.java:
```diff
@@ -0,0 +1,33 @@
+public class ScyllaStoreManager extends CQLStoreManager {
+
+    public ScyllaStoreManager(Configuration configuration) throws BackendException {
+        this(configuration, DEFAULT_MUTATE_MANY_FUNCTION_BUILDER, DEFAULT_STORE_FEATURES_BUILDER, DEFAULT_CQL_SESSION_BUILDER, DEFAULT_PROGRAMMATIC_CONFIGURATION_LOADER_BUILDER);
```
Haven't tested it, but wouldn't this also already work as that has exactly the same implementation?
```suggestion
        super(configuration);
```

FlorianHockmann commented on janusgraph-cql/src/test/java/org/janusgraph/JanusGraphCassandraContainer.java:
```diff
@@ -140,6 +155,11 @@ public JanusGraphCassandraContainer(boolean bindDefaultPort) {
     }
 
+    private void applyCassandraServerConfiguration(String resourcePath){
+        withClasspathResourceMapping(resourcePath, "/etc/cassandra/cassandra.yaml", BindMode.READ_WRITE);
```
Just for my own understanding: These two lines will mount the config file into `/etc/cassandra/cassandra.yaml` AND `/etc/scylla/scylla.yaml` for the container, irrespective of whether it's Cassandra or Scylla, so each service can use its config file in the expected location, right?
And you probably did it like this because you wanted to avoid the complexity of having to distinguish again between Scylla and Cassandra only to save one more mount that isn't necessary?

If yes, then we can of course keep it like this. I'm really just asking to better understand this. (But maybe a small comment would help here in case anyone else later wonders about this.)

porunov replied:
Yes, initially I just decided to add the same config to both places for Cassandra and for Scylla even so one will be redundant because the `JanusGraphCassandraContainer` is for testing purposes, so it shouldn't be a problem to have redundant configs there.
That said, as you mentioned it I realized that it makes sense to simply reuse `isScylla()` method here and mount only one configuration depending on the container type. I will change this method here.

FlorianHockmann reviewed (commented):
Thanks a lot for tackling this, @porunov! I also tried that a while back, but then stopped when I ran into the issue you're describing in #3580 as I had no idea how to resolve this. But your approach of first only adding this as a module and resolve the problem with selecting the right library in the distribution at runtime in a follow-up issue definitely makes sense to get us moving here :-)

I just reviewed the diff here and everything looks good basically. I only added some small nitpick comments. The logs of the added GH Actions also show that the Scylla driver is really being used there:

```
Using Scylla optimized driver!!!
```

I would like to also build the distribution and try out the approach you've documented with manually replacing the JARs if I find some time in the next few days but wanted to already provide my review comments so far.

porunov commented:
Thanks @FlorianHockmann for the review! I will experiment tomorrow or on Wednesday with ScyllaDB parameters suggested here https://github.com/JanusGraph/janusgraph/issues/2505#issuecomment-1428379299 and update the PR with your suggestions and probably modify the JanusGraphCassandraContainer a bit to use those parameters.

I don't personally like the manual approach of replacing libs and also think we should handle #3580 before we release 1.0.0, but thought it makes sense to ship Scylla module first.

porunov commented:
@FlorianHockmann I fixed the comments you left. I also fixed Hadoop tests. 
That said, I noticed that we didn't apply configuration for ScyllaDB. After fixing that I see that `ssl` tests are failing. The reason is described here: https://github.com/JanusGraph/janusgraph/issues/2505#issuecomment-1429165384 
Couldn't figure out the cause why ScyllaDB is failing to start yet :/

porunov commented:
I'm stuck with the issue about ssl certificates ([here](https://github.com/JanusGraph/janusgraph/issues/2505#issuecomment-1431330743)). I'm not sure I'll be able to resolve that issue. Thus, I can't just comment out tests related to ScyllaDB ssl certificates and open a new issue for the tests. Let me know what you think about it.

porunov commented:
Commented the failing tests. The PR is ready for re-review
